src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,712 @@
+    /// is used to populate the Product<C> struct. The Product<C> struct is
+    /// added to the Lookup and finally returned by the method.
+    pub(in crate::plonk) fn commit_product<
+        C: CurveAffine,
+        HBase: Hasher<C::Base>,
+        HScalar: Hasher<C::Scalar>,
+    >(
+        &self,
+        permuted: &Permuted<C>,
+        pk: &ProvingKey<C>,
+        params: &Params<C>,
+        theta: ChallengeTheta<C::Scalar>,
+        beta: ChallengeBeta<C::Scalar>,
+        gamma: ChallengeGamma<C::Scalar>,
+        advice_values: &[Polynomial<C::Scalar, LagrangeCoeff>],
+        fixed_values: &[Polynomial<C::Scalar, LagrangeCoeff>],
+        aux_values: &[Polynomial<C::Scalar, LagrangeCoeff>],
+        transcript: &mut Transcript<C, HBase, HScalar>,
+    ) -> Result<Product<C>, Error> {
+        let permuted = permuted.clone();
+        let unpermuted_input_values: Vec<Polynomial<C::Scalar, LagrangeCoeff>> = self
+            .input_columns
+            .iter()
+            .map(|&input| match input.column_type() {
+                Any::Advice => advice_values[input.index()].clone(),
+                Any::Fixed => fixed_values[input.index()].clone(),
+                Any::Aux => aux_values[input.index()].clone(),
+            })
+            .collect();
+
+        let unpermuted_table_values: Vec<Polynomial<C::Scalar, LagrangeCoeff>> = self
+            .table_columns
+            .iter()
+            .map(|&table| match table.column_type() {
+                Any::Advice => advice_values[table.index()].clone(),
+                Any::Fixed => fixed_values[table.index()].clone(),
+                Any::Aux => aux_values[table.index()].clone(),
+            })
+            .collect();
+
+        // Goal is to compute the products of fractions
+        //
+        // (a_1(\omega^i) + \theta a_2(\omega^i) + ... + beta)(s_1(\omega^i) + \theta(\omega^i) + ... + \gamma) /
+        // (a'(\omega^i) + \beta)(s'(\omega^i) + \gamma)
+        //
+        // where a_j(X) is the jth input column in this lookup,
+        // where a'(X) is the compression of the permuted input columns,
+        // s_j(X) is the jth table column in this lookup,
+        // s'(X) is the compression of the permuted table columns,
+        // and i is the ith row of the column.
+        let mut lookup_product = vec![C::Scalar::one(); params.n as usize];
+
+        // Denominator uses the permuted input column and permuted table column
+        parallelize(&mut lookup_product, |lookup_product, start| {
+            for ((lookup_product, permuted_input_value), permuted_table_value) in lookup_product
+                .iter_mut()
+                .zip(permuted.permuted_input_value[start..].iter())
+                .zip(permuted.permuted_table_value[start..].iter())
+            {
+                *lookup_product *= &(*beta + permuted_input_value);
+                *lookup_product *= &(*gamma + permuted_table_value);
+            }
+        });
```

Comment:
This would be more efficient (half the number of multiplications):
```suggestion
        let mut lookup_product = vec![C::Scalar::zero(); params.n as usize];

        // Denominator uses the permuted input column and permuted table column
        parallelize(&mut lookup_product, |lookup_product, start| {
            for ((lookup_product, permuted_input_value), permuted_table_value) in lookup_product
                .iter_mut()
                .zip(permuted.permuted_input_value[start..].iter())
                .zip(permuted.permuted_table_value[start..].iter())
            {
                *lookup_product = (*beta + permuted_input_value) * &(*gamma + permuted_table_value);
            }
        });
```

Reply:
Nice, note that some of these optimizations also apply to the permutation argument probably so it'd be nice to mirror them over there as well.